Skill for completing tech breakdowns (#148)
bitwarden/ai-plugins: plugins/bitwarden-delivery-tools/skills/completing-breakdown/SKILL.md

---
name: completing-breakdown

> description: Complete a Bitwarden Tech Breakdown — update the Status block to `Complete` and move the breakdown folder into the team's `complete/` subfolder so the active directory only contains in-flight work. Use when a team has finished delivering the work captured in a Tech Breakdown and is ready to archive it. Triggered by phrasings such as "complete this breakdown", "mark the breakdown as done", "archive the breakdown", "move breakdown to complete", "finish the breakdown", "wrap up the breakdown".

Reviewer: 🎨 I don't think that second block of text in sentence 1 is needed, we should use completed instead of finished, and trim down the second sentence.

```suggestion
description: Complete a Bitwarden Tech Breakdown. Use when a team has completed delivering the work captured in a Tech Breakdown. Triggered by phrasings such as "complete this breakdown", "mark the breakdown as done", "archive the breakdown", "move breakdown to complete", "finish the breakdown", "wrap up the breakdown".
```

Author: Addressed in [`f423f4a` (this PR)](https://github.com/bitwarden/ai-plugins/pull/148/commits/f423f4aebade7b12450bf1e3a4feca037862f36e).

argument-hint: "[<breakdown-path | jira-key | slug>]"
arguments: breakdown
allowed-tools: Read, Edit, Glob, Bash(git mv:*), Bash(git status:*), Bash(mkdir:*), AskUserQuestion
---

# Completing a Tech Breakdown

## Overview

> Help the user retire a Tech Breakdown once the work it describes has shipped. The skill flips the Status block in `breakdown.md` to `Complete`, then moves the entire breakdown folder (`breakdown.md`, `tasks.md`, any sibling artifacts) under the team's `complete/` subdirectory. The active team directory should only contain breakdowns that are still in flight; the `complete/` subfolder is the archive that preserves the design record without cluttering navigation, while keeping it discoverable for in-flight research.

Reviewer: ♻️ / 🎨 The completion process is very simple; I recommend refraining from repeating it here & follow K.I.S.S.

```suggestion
Help the user retire a Tech Breakdown once the work it describes has shipped. 
```

Author: Addressed in [`f423f4a` (this PR)](https://github.com/bitwarden/ai-plugins/pull/148/commits/f423f4aebade7b12450bf1e3a4feca037862f36e).

- **Don't touch sibling artifacts.** `tasks.md` and other files in the folder move with the folder unchanged.

## Phases

### Phase 1: Resolve the breakdown

Use the orientation rules from the HARD-GATE to locate `<team>/<JIRA-KEY>-<short-slug>/breakdown.md` in the `tech-breakdowns` working copy. Record the resolved values:

- `BREAKDOWN_PATH` — absolute path to `breakdown.md`.
- `BREAKDOWN_FOLDER` — the parent folder.
- `TEAM_DIR` — the team directory (`BREAKDOWN_FOLDER`'s parent).
- `DEST_FOLDER` — `<TEAM_DIR>/complete/<JIRA-KEY>-<short-slug>/`.

Read `breakdown.md` and surface the current Status block to the user before proceeding.

### Phase 2: Update the Status block

Edit `breakdown.md` in place. Update the Status block:

- `Status:` — `Complete`
- `Last substantive update:` — today's date + the literal note `breakdown completed`

Leave `Active owner / contact:` unchanged; the historical record should keep pointing at the engineer who ran the breakdown.

Do not edit the Spec, Plan, Tasks, or any other section. The breakdown's content is now the as-shipped record.

### Phase 3: Move the folder

> 1. Ensure the `complete/` directory exists under `TEAM_DIR`. `git mv` does not create missing parent directories — if `<TEAM_DIR>/complete/` does not exist yet, run `mkdir -p <TEAM_DIR>/complete/` first. The first breakdown a team completes will hit this case.

Reviewer: ⚠️ / ♻️ A local multi-agent review had the following finding. Claude was overly verbose, but I think the sentiment that it's getting at is fair to consider. 

<details><summary>Resolved path values interpolated into `mkdir`/`git mv` without the validation guard the sibling skill mandates (CWE-78/CWE-22)</summary>

<br/> `plugins/bitwarden-delivery-tools/skills/completing-breakdown/SKILL.md:58-59`
<br/> **Caught by:** Code quality agent + Security & logic agent
<br/> **Caught by model(s):** both (opus + sonnet) — **rated differently:** opus ⚠️ Important, sonnet ♻️ Refactor
<br/>  ✅ HOLDS — confirmed: Phase 3 interpolates `TEAM_DIR`/`BREAKDOWN_FOLDER`/`DEST_FOLDER` into unquoted `mkdir -p`/`git mv` with no validation; `starting-breakdown:51` mandates the opposite verbatim ("never interpolate an non-validated value into `mkdir`, `cp`, or any other shell command"). cmon-bro empirically confirmed an unquoted path containing a space word-splits and breaks the move. Severity is the genuine judgment call; see note.

<br/> Phase 3 interpolates `TEAM_DIR`, `BREAKDOWN_FOLDER`, and `DEST_FOLDER` — derived from the `$breakdown` argument (a path/Jira-key/slug the operator types) and from `Glob`-discovered folder names — into `mkdir -p <TEAM_DIR>/complete/` and `git mv <BREAKDOWN_FOLDER> <DEST_FOLDER>` with no validation and no quoting. The sibling `starting-breakdown:51` establishes the opposite convention as a hard rule for this exact namespace: slug must match `^[a-z][a-z0-9-]*$`, Jira key must match `^[A-Z][A-Z0-9]+-[0-9]+$`, and "never interpolate an non-validated value into `mkdir`, `cp`, or any other shell command." The `allowed-tools` wildcards (`Bash(mkdir:*)`, `Bash(git mv:*)`) gate only the command prefix, not metacharacter-driven argument injection in the operands.

**Severity split (cmon-bro adjudication):** opus rated this Important; sonnet's audit downgraded to Refactor on the grounds that the target is design-record markdown in a first-party docs repo (no Vault Data / EK / AT / SC boundary), so the realistic worst case is workflow disruption, not a breach. cmon-bro leans **Important**: the sibling treats this guard as a hard security control the team already chose to enforce in the same namespace, dropping it reintroduces that class, and the failure mode is not purely hypothetical — a `$breakdown` path or folder name containing a space or shell metacharacter reaches an unquoted command (reproduced: an unquoted path with a space word-splits and corrupts the `git mv` argument list). Reasonable reviewers could land on Refactor given that well-formed breakdown folders are slug/key-validated at creation time.

**Suggested fix:** add the validation clause `starting-breakdown` uses — validate the resolved Jira key and slug against the documented regexes, reject any resolved path containing `..` or characters outside `[A-Za-z0-9._/-]`, single-quote every interpolated path, and confirm the resolved path stays within the `tech-breakdowns` working copy. Implicates **P05** and the **TC** boundary.

  </details>

Author: Addressed in [`f423f4a` (this PR)](https://github.com/bitwarden/ai-plugins/pull/148/commits/f423f4aebade7b12450bf1e3a4feca037862f36e).

2. Run `git mv <BREAKDOWN_FOLDER> <DEST_FOLDER>` from inside the `tech-breakdowns` working copy. This moves `breakdown.md`, `tasks.md`, and any sibling artifacts in one operation while preserving history.

> 3. Surface the result to the user:

Reviewer: ♻️ / 🎨 On line 29 we instruct Claude to 

> Surface the source path, destination path, and the affected files once before doing the move.

And then in Phase 3 we instruct Claude to do the opposite; move the tech breakdown first and then tell the human. I would consider reordering one of these instructions.

Author: Addressed in [`f423f4a` (this PR)](https://github.com/bitwarden/ai-plugins/pull/148/commits/f423f4aebade7b12450bf1e3a4feca037862f36e).
